### DISCUSSION TITLE
Select audio/ Prefer language

### DISCUSSION THREAD

**Discussion by muisje** [McCloudS/subgen#149] [General]
So let's say there's a video with audio in multiple languages. And i would only be interested in one language. In my case Spanish. So i would like subgen to choose the Spanish audio track if available. (If not just skip it). 

What happens is that subgen takes the English audio track. And then doesn't generate any subtitles because i configured it to skip English.

```
> mkvinfo movie.mkv

|+ Tracks
| + Track
|  + Track number: 1 (track ID for mkvmerge & mkvextract: 0)
|  + Track UID: 14974003125132585736
|  + Track type: video
|  + "Lacing" flag: 0
|  + Language: spa
|  + Codec ID: V_MPEG4/ISO/AVC
|  + Codec's private data: size 46 (H.264 profile: High @L4.1)
|  + Default duration: 00:00:00.041708333 (23.976 frames/fields per second for a video track)
|  + Language (IETF BCP 47): es
|  + Video track
|   + Pixel width: 1920
|   + Pixel height: 802
|   + Display width: 1920
|   + Display height: 802
|   + Video color information
|    + Color matrix coefficients: 1
|    + Horizontal chroma siting: 1
|    + Vertical chroma siting: 2
|    + Color range: 1
|    + Color transfer: 1
|    + Color primaries: 1
| + Track
|  + Track number: 2 (track ID for mkvmerge & mkvextract: 1)
|  + Track UID: 7766254991514104971
|  + Track type: audio
|  + "Forced display" flag: 1
|  + Language: spa
|  + Codec ID: A_AC3
|  + Default duration: 00:00:00.032000000 (31.250 frames/fields per second for a video track)
|  + Language (IETF BCP 47): es
|  + Audio track
|   + Sampling frequency: 48000
|   + Channels: 6
| + Track
|  + Track number: 3 (track ID for mkvmerge & mkvextract: 2)
|  + Track UID: 10288278466344406720
|  + Track type: audio
|  + "Default track" flag: 0
|  + Codec ID: A_AC3
|  + Default duration: 00:00:00.032000000 (31.250 frames/fields per second for a video track)
|  + Language (IETF BCP 47): en
|  + Audio track
|   + Sampling frequency: 48000
|   + Channels: 6
| + Track
|  + Track number: 4 (track ID for mkvmerge & mkvextract: 3)
|  + Track UID: 4079945488710420686
|  + Track type: audio
|  + "Default track" flag: 0
|  + Codec ID: A_DTS
|  + Default duration: 00:00:00.010666667 (93.750 frames/fields per second for a video track)
|  + Language (IETF BCP 47): en
|  + Audio track
|   + Sampling frequency: 48000
|   + Channels: 6
|   + Bit depth: 24
```

output when it get's triggered by a jellyfin webhook:

`Language 'eng' detected in movie.mkv and is in the skip list ['eng'], skipping subtitle generation`

Which results in no subtitles even though there's a spanish audio track that would like to be transcribed. 

With these settings i would imagine nothing would be skipped, because nothing will be English. But i guess forced detect language will not be used when checking the language with the sample.  
       - "SKIP_LANG_CODES=eng"
       - "FORCE_DETECTED_LANGUAGE_TO=es"

My proposal is:
*  to have subgen check if there are other audios that are not meant to be skipped. 
*  Have an option to only transcribe for a specific language. (Which is my use case)
*  Have an option to either transcribe the default language of the video or preferred language(s) or all

With these options the users of subgen i would think make sure they get the right subtitles transcribed.

**Comment by McCloudS** [comment]
See https://www.reddit.com/r/selfhosted/s/Z5wLmXq3qu for a semi-related discussion.  The default behavior of stable-ts and pyav is dictating which audio track is picked, I haven’t done any additional work or thought what it might look like to sift through the streams and encode individual languages.  I’m not even sure if it’s a problem worth solving.

I’ll think on it.

**Comment by McCloudS** [comment]
I believe bazarr is also a workaround for this as it tracks the languages of audio and subtitles.

**Comment by muisje** [comment]
I didn't have the issue in mentioned in the reddit comment, but i guess this would solve according to that theory. It now extracts one audio track if it has multiple in the pull request #144 . Bazarr didn't work well because it was using detect language which is not totally reliable and lacks control. I think my pull request solves the issue. I tested it with TRANSCRIBE_FOLDERS set. And the folder with a couple files of http://cdn.media.ccc.de/congress/2019/h264-hd/ which have videos with multiple and single audio tracks.

**Comment by muisje** [comment]
So first of all sorry for breaking this repository 🫣.  I updated the code some more. Now handles multiple preferred languages and option to limit subtitle generation to it. 

It also now has config options to not skip or not on specific conditions, like if audio language is known or not, if it only should check external subgen generated subtitles. 

 Besides that i changed some problematic code that was causing unwanted behavior. Related to LanguaCode. And changed the code in has_audio to better exclude non audio having files that could be for some reason in the folder like txt files besides images. 

I tested on my own library besides some samples. There i get an empty error from av in has_subtitle_language_in_file(). Not sure why.

https://github.com/McCloudS/subgen/compare/main...muisje:subgen:LanguageCode-improvements

**Comment by McCloudS** [comment]
No worries, I've done it a dozen or more times.

I'll take a look at the rest.  The av stuff has been rock solid for over a year, so if you have some info, we can take a look.  I've previously left detect-language out of the queue because it didn't seem worth an effort that should run for mere seconds.

**Comment by muisje** [comment]
Hmmm maybe add the detect language in the bottom of the queue in the case we do not know the language. Now it just adds them to the queue when it recognizes it couldn't determine the language to transcribe, but maybe we can have like a priority in the queue?

**Comment by muisje** [comment]
uhm  the error i get from pyav is empty

mkvinfo on the file:
```

+ EBML head
|+ EBML version: 1
|+ EBML read version: 1
|+ Maximum EBML ID length: 4
|+ Maximum EBML size length: 8
|+ Document type: matroska
|+ Document type version: 4
|+ Document type read version: 2
+ Segment: size 3781213426
|+ Seek head (subentries will be skipped)
|+ EBML void: size 4012
|+ Segment information
| + Timestamp scale: 1000000
| + Multiplexing application: libebml v1.3.9 + libmatroska v1.5.2
| + Writing application: mkvmerge v35.0.0 ('All The Love In The World') 64-bit
| + Duration: 01:41:10.112000000
| + Date: 2020-04-18 21:47:06 UTC
| + Title: Movie
| + Segment UID: 0x92 0x31 0x75 0xaa 0x2f 0x4d 0x15 0x7b 0x35 0xf1 0xf1 0x8d 0xd3 0xf6 0x7a 0xd9
|+ Tracks
| + Track
|  + Track number: 1 (track ID for mkvmerge & mkvextract: 0)
|  + Track UID: 13112506145182749843
|  + Track type: video
|  + "Lacing" flag: 0
|  + Codec ID: V_MPEG4/ISO/AVC
|  + Codec's private data: size 43 (H.264 profile: High @L4.0)
|  + Default duration: 00:00:00.041708333 (23.976 frames/fields per second for a video track)
|  + Video track
|   + Pixel width: 1908
|   + Pixel height: 792
|   + Display width: 1908
|   + Display height: 792
| + Track
|  + Track number: 2 (track ID for mkvmerge & mkvextract: 1)
|  + Track UID: 12710499002606413153
|  + Track type: audio
|  + "Forced display" flag: 1
|  + Codec ID: A_AC3
|  + Default duration: 00:00:00.032000000 (31.250 frames/fields per second for a video track)
|  + Language: spa
|  + Name: Spanish
|  + Audio track
|   + Sampling frequency: 48000
|   + Channels: 6
|+ EBML void: size 1103
|+ Chapters
| + Edition entry
|  + Edition flag hidden: 0
|  + Edition flag default: 0
|  + Edition UID: 14774111180023649384
|  + Chapter atom
|   + Chapter UID: 1846176818477802661
|   + Chapter time start: 00:00:00.000000000
|   + Chapter flag hidden: 0
|   + Chapter flag enabled: 1
|   + Chapter display
|    + Chapter string: 1
|    + Chapter language: eng
|  + Chapter atom
|   + Chapter UID: 15214367168180763248
|   + Chapter time start: 00:02:37.000000000
|   + Chapter flag hidden: 0
|   + Chapter flag enabled: 1
|   + Chapter display
|    + Chapter string: 2
|    + Chapter language: eng
|  + Chapter atom
|   + Chapter UID: 6149532507709372122
|   + Chapter time start: 00:11:31.000000000
|   + Chapter flag hidden: 0
|   + Chapter flag enabled: 1
|   + Chapter display
|    + Chapter string: 3
|    + Chapter language: eng
|  + Chapter atom
|   + Chapter UID: 2465782857638220241
|   + Chapter time start: 00:17:44.000000000
|   + Chapter flag hidden: 0
|   + Chapter flag enabled: 1
|   + Chapter display
|    + Chapter string: 4
|    + Chapter language: eng
|  + Chapter atom
|   + Chapter UID: 3065729896823724741
|   + Chapter time start: 00:27:08.000000000
|   + Chapter flag hidden: 0
|   + Chapter flag enabled: 1
|   + Chapter display
|    + Chapter string: 5
|    + Chapter language: eng
|  + Chapter atom
|   + Chapter UID: 7259181674930921508
|   + Chapter time start: 00:38:05.000000000
|   + Chapter flag hidden: 0
|   + Chapter flag enabled: 1
|   + Chapter display
|    + Chapter string: 6
|    + Chapter language: eng
|  + Chapter atom
|   + Chapter UID: 3808542860049580512
|   + Chapter time start: 00:47:32.000000000
|   + Chapter flag hidden: 0
|   + Chapter flag enabled: 1
|   + Chapter display
|    + Chapter string: 7
|    + Chapter language: eng
|  + Chapter atom
|   + Chapter UID: 13641716792480369630
|   + Chapter time start: 00:57:44.000000000
|   + Chapter flag hidden: 0
|   + Chapter flag enabled: 1
|   + Chapter display
|    + Chapter string: 8
|    + Chapter language: eng
|  + Chapter atom
|   + Chapter UID: 42503990203636808
|   + Chapter time start: 01:06:33.000000000
|   + Chapter flag hidden: 0
|   + Chapter flag enabled: 1
|   + Chapter display
|    + Chapter string: 9
|    + Chapter language: eng
|  + Chapter atom
|   + Chapter UID: 1406997278544134674
|   + Chapter time start: 01:15:55.000000000
|   + Chapter flag hidden: 0
|   + Chapter flag enabled: 1
|   + Chapter display
|    + Chapter string: 10
|    + Chapter language: eng
|  + Chapter atom
|   + Chapter UID: 14061341534077077676
|   + Chapter time start: 01:25:03.000000000
|   + Chapter flag hidden: 0
|   + Chapter flag enabled: 1
|   + Chapter display
|    + Chapter string: 11
|    + Chapter language: eng
|  + Chapter atom
|   + Chapter UID: 5656980010924663617
|   + Chapter time start: 01:37:18.000000000
|   + Chapter flag hidden: 0
|   + Chapter flag enabled: 1
|   + Chapter display
|    + Chapter string: 12
|    + Chapter language: eng
|+ EBML void: size 101
|+ Cluster
```

**Comment by McCloudS** [comment]
Made some edits, I want to eventually fix detect_language (encode: true) to use some of your audio language track logic.  It's kind of broken now and seems to do all audio at the same time.  I defaulted user inputs/env variables to to_string.  I think it will combat user input errors and prevent some settings being 2 letter codes and some 3.  

https://github.com/McCloudS/subgen/blob/LanguageCodeImprovements/subgen.py

**Comment by muisje** [comment]
In which part is that happening? Detect language for bazarr or when it gets triggered when it couldn't determine the language when it wants to generate subtitles?

**Comment by McCloudS** [comment]
Made some small edits, essentially trying to get ASR and detect-language to be able to take any file, like an MKV or mp4 and actually work.  I get wonky results on both.  encode=true is essentially a passthrough since faster-whisper handles the audio file, but I get this error on /asr:
```
2024-12-03 20:27:18,648 INFO: Transcribing file 'output.mp4' from Bazarr/ASR webhook
2024-12-03 20:27:18,648 DEBUG: Model was purged, need to re-create
2024-12-03 20:27:19,098 INFO: Task Bazarr-asr-1w7aIb is being handled by ASR.
2024-12-03 20:27:19,098 DEBUG: There are 0 tasks left in the queue.
2024-12-03 20:27:20,294 INFO: Processing audio with duration 07:17.984
2024-12-03 20:27:20,568 INFO: Detected language 'en' with probability 0.91
Detected Language: english
Transcribe:  92%|██████████████████████████████████████████████████████████████████▉      | 401.32/437.98 [00:37<00:01, 21.84sec/s]2024-12-03 20:27:58,225 DEBUG: No speech threshold is met (0.927697 > 0.600000)
Transcribe: 100%|█████████████████████████████████████████████████████████████████████████| 437.98/437.98 [00:37<00:00, 11.63sec/s]
[mov,mp4,m4a,3gp,3g2,mj2 @ 0x15ae05ff0] stream 3, offset 0x34: partial file
[mov,mp4,m4a,3gp,3g2,mj2 @ 0x15ae05ff0] stream 1, offset 0x4aeea: partial file
[in#0/mov,mp4,m4a,3gp,3g2,mj2 @ 0x600001e0c000] Error during demuxing: Invalid data found when processing input
2024-12-03 20:28:00,024 ERROR: Error processing or transcribing Bazarr file: output.mp4 -- Exception: FFmpeg failed to load audio from bytes (484719474).
2024-12-03 20:28:00,085 DEBUG: Queue is empty, clearing/releasing VRAM
INFO:     127.0.0.1:57279 - "POST /asr?task=transcribe&video_file=output.mp4&encode=true&output=srt&word_timestamps=false HTTP/1.1" 200 OK

```
and I think detect-language has an issue with: `whisper.pad_or_trim(audio_file.file.read() , sample_rate * int(detect_language_length))` using the same file with encode:
```
2024-12-03 20:33:27,729 DEBUG: Model was purged, need to re-create
2024-12-03 20:33:28,365 INFO: Task Bazarr-detect-language-74UhOE is being handled by ASR.
2024-12-03 20:33:28,365 DEBUG: There are 0 tasks left in the queue.
2024-12-03 20:33:28,700 INFO: Error processing or transcribing Bazarr output.mp4: 'bytes' object has no attribute 'shape'
2024-12-03 20:33:28,761 DEBUG: Queue is empty, clearing/releasing VRAM
INFO:     127.0.0.1:57297 - "POST /detect-language?encode=true&detect_lang_length=30&detect_lang_offset=90 HTTP/1.1" 200 OK
```
Probably because the pad_or_trim expects a numpy array.

**Comment by McCloudS** [comment]
Actually the ASR error appears to be on my end, I have a bad container when I split a stream from an mkv to mp4.  Using the original files have no issue.  

Just need to figure out the detect-language part.  Probably just need to use your function to return a file in memory with the offset and length.

**Comment by muisje** [comment]
With extract_audio_segment_to_memory you already have like a small part of the audio. Don't think you need use a function like pad_or_trim on it. Without it I think it should work.

Sent from Outlook for Android<https://aka.ms/AAb9ysg>
________________________________

**Comment by McCloudS** [comment]
Merged your changes with some modifications from me.  Nothing seems broken yet.
